### PR TITLE
Reimplement pick16SCandidates in a fixed-memory fashion

### PR DESCRIPTION
**Pull request by laughedelic**
Using `Iterable` instead of `Stream` and do some lazy `groupBy`.

### CODE REVIEW

File: src/test/scala/pick16SCandidates.scala [hunk 1]
```diff
@@ -104,19 +104,44 @@ case object pick16SCandidates extends FilterData(
...
+
+      // NOTE: Buffered iterator has .head which doesn't pop it as .next
+      private var rest: BufferedIterator[V] = iterator.buffered
+
+      def hasNext: Boolean = rest.hasNext
+
+      def next(): (K, Seq[V]) = {
+
+        val key = getKey(rest.head)
+        val (prefix, suffix) = rest.span { getKey(_) == key }
```
Comment (eparejatobes) [inline src/test/scala/pick16SCandidates.scala]: span leaks, not safe.

**Comment by laughedelic** [comment]
@eparejatobes please, review 5a80b98

**Comment by eparejatobes** [comment]
OK, will do.

File: src/test/scala/pick16SCandidates.scala [hunk 2]
```diff
@@ -104,19 +104,49 @@ case object pick16SCandidates extends FilterData(
...
+      private val rest: BufferedIterator[V] = iterator.buffered
+
+      @annotation.tailrec
+      private def longestPrefix(key: K, acc: Seq[V]): Seq[V] = {
+        if ( rest.hasNext && getKey(rest.head) == key )
+          longestPrefix(key, rest.next() +: acc)
+        else acc
+      }
+
+      def hasNext: Boolean = rest.hasNext
```
Comment (eparejatobes) [inline src/test/scala/pick16SCandidates.scala:119]: document that this is OK because of the contiguous grouping assumpltion

Comment (laughedelic) [reply src/test/scala/pick16SCandidates.scala:119]: This is written at the line 110, but I can add a note here too.

File: src/test/scala/pick16SCandidates.scala [hunk 3]
```diff
@@ -104,19 +104,49 @@ case object pick16SCandidates extends FilterData(
...
 
+  // TODO: move it to db.rnacentral
+  implicit class IteratorOps[V](val iterator: Iterator[V]) extends AnyVal {
+
+    /* Similar to the Stream's .groupBy, but assuming that groups are contiguous. Another difference is that it returns the key corresponding to each group. */
+    // NOTE: The original iterator should be discarded after calling this method
+    def groupBy[K](getKey: V => K): Iterator[(K, Seq[V])] = new Iterator[(K, Seq[V])] {
+      /* The definition is very straightforward: we keep the `rest` of values and on each `.next()` call bite off the longest prefix with the same key */
+
+      /* Buffered iterator allows to loo ahead without removing the next element */
```
Comment (eparejatobes) [inline src/test/scala/pick16SCandidates.scala]: *look

File: src/test/scala/pick16SCandidates.scala [hunk 4]
```diff
@@ -104,19 +104,49 @@ case object pick16SCandidates extends FilterData(
...
+      private val rest: BufferedIterator[V] = iterator.buffered
```
Comment (eparejatobes) [inline src/test/scala/pick16SCandidates.scala:116]: better for this to be lazy?

Comment (laughedelic) [reply src/test/scala/pick16SCandidates.scala:116]: Why? Once you called the method, you just construct this buffered iterator. It doesn't get evaluated or anything, so what's the point?

File: src/test/scala/pick16SCandidates.scala [hunk 5]
```diff
@@ -104,19 +104,49 @@ case object pick16SCandidates extends FilterData(
...
+
+      @annotation.tailrec
+      private def longestPrefix(key: K, acc: Seq[V]): Seq[V] = {
```
Comment (eparejatobes) [inline src/test/scala/pick16SCandidates.scala]: why prefix here? unless I'm missing something, this is not a prefix of anything.

Comment (laughedelic) [reply src/test/scala/pick16SCandidates.scala]: It's a prefix of the `rest` iterator. Longest contiguous group of elements that are mapped to the same `key`

File: src/test/scala/pick16SCandidates.scala (hunk 4 above)
Comment (eparejatobes) [reply src/test/scala/pick16SCandidates.scala:116]: I don't know how much stuff is involved in just _creating_ a buffered iterator

File: src/test/scala/pick16SCandidates.scala (hunk 5 above)
Comment (eparejatobes) [reply src/test/scala/pick16SCandidates.scala]: OK, fair enough. I wouldn't think about it that way, but it's fine.

File: src/test/scala/pick16SCandidates.scala (hunk 4 above)
Comment (laughedelic) [reply src/test/scala/pick16SCandidates.scala:116]: [Not much](https://github.com/scala/scala/blob/v2.11.8/src/library/scala/collection/Iterator.scala#L991-L1011). Do you really think it should be lazy? I just don't see the point.

File: src/test/scala/pick16SCandidates.scala (hunk 5 above)
Comment (eparejatobes) [reply src/test/scala/pick16SCandidates.scala]: See 2cd551c

File: src/test/scala/pick16SCandidates.scala [hunk 6]
```diff
@@ -115,21 +115,23 @@ case object pick16SCandidates extends FilterData(
...
       }
+
+      @annotation.tailrec
+      private def groupOf_rec(key: K, acc: Seq[V]): Seq[V] = {
+        if ( rest.hasNext && getKey(rest.head) == key )
+          groupOf_rec(key, rest.next() +: acc)
+        else acc
+      }
+
+      private def groupOf(key: K): Seq[V] = groupOf_rec(key, Seq())
```
Comment (laughedelic) [inline src/test/scala/pick16SCandidates.scala:134]: To be honest, I don't think that this code (neither the name, nor the order) looks more clear. And I don't see any point to add `groupOf` method to use it once. But anyway, it's fine 👍

**Comment by laughedelic** [comment]
Anything else here?

**Comment by eparejatobes** [comment]
No, fine for me. LGTM

**Pull request merged**